test/common/websocket/codec_test.cc

```diff
@@ -0,0 +1,824 @@
+
+  ASSERT_TRUE(frames.has_value());
+  EXPECT_EQ(true, frames.value()[0].masking_key_.has_value());
+  EXPECT_EQ(0x4f5a12e2, frames.value()[0].masking_key_.value());
+  EXPECT_EQ(1, frames.value().size());
+  EXPECT_EQ(0, frames.value()[0].payload_length_);
+}
+
+TEST(WebSocketCodecTest, decodeFramesWithNoPayloadNonMasked7BitLength) {
+  Buffer::OwnedImpl buffer;
+  Buffer::addSeq(buffer, {0x81, 0x00});
+
+  Decoder decoder;
+  absl::optional<std::vector<Frame>> frames = decoder.decode(buffer);
+
+  ASSERT_TRUE(frames.has_value());
+  EXPECT_EQ(false, frames.value()[0].masking_key_.has_value());
+  EXPECT_EQ(1, frames.value().size());
+  EXPECT_EQ(0, frames.value()[0].payload_length_);
+}
+
+TEST(WebSocketCodecTest, decodeFramesWithNoPayloadMasked16BitLength) {
+  Buffer::OwnedImpl buffer;
+  Buffer::addSeq(buffer, {0x81, 0xfe, 0x00, 0x00, 0x3c, 0x33, 0x2a, 0x16});
+
+  Decoder decoder;
+  absl::optional<std::vector<Frame>> frames = decoder.decode(buffer);
+
+  ASSERT_TRUE(frames.has_value());
+  EXPECT_EQ(1, frames.value().size());
+  EXPECT_EQ(true, frames.value()[0].masking_key_.has_value());
+  EXPECT_EQ(0x3c332a16, frames.value()[0].masking_key_.value());
+  EXPECT_EQ(1, frames.value().size());
+  EXPECT_EQ(0, frames.value()[0].payload_length_);
+}
+
+TEST(WebSocketCodecTest, decodeFramesWithNoPayloadNonMasked16BitLength) {
+  Buffer::OwnedImpl buffer;
+  Buffer::addSeq(buffer, {
+                             0x81,
+                             0x7e,
+                             0x00,
+                             0x00,
+                         });
+
+  Decoder decoder;
+  absl::optional<std::vector<Frame>> frames = decoder.decode(buffer);
+
+  ASSERT_TRUE(frames.has_value());
+  EXPECT_EQ(false, frames.value()[0].masking_key_.has_value());
+  EXPECT_EQ(1, frames.value().size());
+  EXPECT_EQ(0, frames.value()[0].payload_length_);
+}
+
+TEST(WebSocketCodecTest, decodeFramesWithNoPayloadMasked64BitLength) {
+  Buffer::OwnedImpl buffer;
+  Buffer::addSeq(
+      buffer, {0x81, 0xff, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x1f, 0xfa, 0x4f, 0x5a});
+
+  Decoder decoder;
+  absl::optional<std::vector<Frame>> frames = decoder.decode(buffer);
+
+  ASSERT_TRUE(frames.has_value());
+  EXPECT_EQ(true, frames.value()[0].masking_key_.has_value());
+  EXPECT_EQ(0x1ffa4f5a, frames.value()[0].masking_key_.value());
+  EXPECT_EQ(1, frames.value().size());
+  EXPECT_EQ(0, frames.value()[0].payload_length_);
+}
+
+TEST(WebSocketCodecTest, decodeFramesWithNoPayloadNonMasked64BitLength) {
+  Buffer::OwnedImpl buffer;
+  Buffer::addSeq(buffer, {0x81, 0x7f, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00});
+
+  Decoder decoder;
+  absl::optional<std::vector<Frame>> frames = decoder.decode(buffer);
+
+  ASSERT_TRUE(frames.has_value());
+  EXPECT_EQ(false, frames.value()[0].masking_key_.has_value());
+  EXPECT_EQ(1, frames.value().size());
+  EXPECT_EQ(0, frames.value()[0].payload_length_);
+}
+
+TEST(WebSocketCodecTest, encodeFrameHeader) {
```

Comment:
Since the Encoder is first in the source file, let's put the encoder tests at the top of the test.cc